### PR TITLE
[VPU] Support for Bidirectional mode in Broadcast DTS

### PR DESCRIPTION
**Pull request by AndrewBakalinIntel**
Ticket - #-40305

### CODE REVIEW

**Comment by AndrewBakalinIntel** [comment]
@gladilov-gleb @andrejsokolov could you please take a look?

**Comment by AndrewBakalinIntel** [comment]
@PoliOwl FYI

**Comment by AndrewBakalinIntel** [comment]
@gladilov-gleb @andrejsokolov ping

**Review by andrejsokolov — APPROVED** [review]

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp [hunk 1]
```diff
@@ -29,19 +31,66 @@ void dynamicToStaticShapeBroadcast(std::shared_ptr<ngraph::Node> target) {
...
                          "provided {}", broadcast->get_broadcast_spec().m_type);
     }
 
-    auto dsr = std::make_shared<ngraph::vpu::op::DynamicShapeResolver>(
-            staticShapeBroadcast->output(0), broadcast->input_value(1));
-    dsr->set_friendly_name(broadcast->get_friendly_name());
+    std::shared_ptr<ngraph::Node> dsr;
+
+    if (broadcast->get_broadcast_spec() == ngraph::op::BroadcastType::BIDIRECTIONAL) {
+        const auto inputShape = broadcast->get_input_partial_shape(0);
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp:48]: Nitpick: you can get static shape by `get_input_shape(0)`

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp [hunk 2]
```diff
@@ -29,19 +31,66 @@ void dynamicToStaticShapeBroadcast(std::shared_ptr<ngraph::Node> target) {
...
+        const auto targetShapeDimsCount = ngraph::shape_size(broadcast->get_input_partial_shape(1).get_shape());
+
+        const auto inputShapeConst = std::make_shared<ngraph::opset5::Constant>(shapeType,
+                                                                                ngraph::Shape{static_cast<size_t>(inputShapeDimsCount)},
+                                                                                inputShape.get_shape());
+
+        const auto maxRankNode = inputShapeDimsCount > targetShapeDimsCount ? inputShapeConst : targetShape;
+        const auto minRankNode = maxRankNode == inputShapeConst ? targetShape : inputShapeConst;
+        const auto maxRank = maxRankNode == inputShapeConst ? inputShapeDimsCount : targetShapeDimsCount;
+        const auto minRank = minRankNode == inputShapeConst ? inputShapeDimsCount : targetShapeDimsCount;
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp:63]:
```suggestion
        const auto minRank = std::min(inputShapeDimsCount, targetShapeDimsCount);
        const auto maxRank = std::max(inputShapeDimsCount, targetShapeDimsCount);
        const auto minRankNode = minRank == inputShapeDimsConst ? inputShapeConst : targetShape;
        const auto maxRankNode = minRank == inputShapeDimsConst ? targetShape : inputShapeConst;
```

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp [hunk 3]
```diff
@@ -29,19 +31,66 @@ void dynamicToStaticShapeBroadcast(std::shared_ptr<ngraph::Node> target) {
...
+
+        ngraph::NodeVector dims;
+        for (int i = 0; i < minRank; i++) {
+            const auto minRankDim = std::make_shared<ngraph::opset5::Gather>(minRankNode,
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {minRank - i - 1}),
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {0}));
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp:69]: Nitpick: such indentation is hard to read

```suggestion
            const auto minRankDim = std::make_shared<ngraph::opset5::Gather>(
                minRankNode,
                ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {minRank - i - 1}),
                ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {0}));
```

Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp:69]: If you agree, apply the same manner for all changes

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp [hunk 4]
```diff
@@ -29,19 +31,66 @@ void dynamicToStaticShapeBroadcast(std::shared_ptr<ngraph::Node> target) {
...
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {maxRank - i - 1}),
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {0}));
+            dims.push_back(std::make_shared<ngraph::opset5::Maximum>(minRankDim, maxRankDim));
+        }
 
+        for (int i = maxRank - minRank - 1; i >= 0; i--) {
+            dims.push_back(std::make_shared<ngraph::opset5::Gather>(maxRankNode,
+                                                                    ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {i}),
+                                                                    ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {0})));
+        }
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp:80]: I think all of this is better to be extracted to a separate helper with `revert` in name - if I got your idea correctly

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp [hunk 5]
```diff
@@ -29,19 +31,66 @@ void dynamicToStaticShapeBroadcast(std::shared_ptr<ngraph::Node> target) {
...
+
+        std::reverse(dims.begin(), dims.end());
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp:82]: Okay, now I'm confused a bit. Code above seems to me as extracting dimensions from shape in rever order, but then you reverse here again. Why is that?

File: inference-engine/tests/functional/plugin/myriad/subgraph_tests/dsr_tests_common.hpp [hunk 6]
```diff
@@ -70,6 +71,8 @@ class DSR_TestsCommon : virtual public LayerTestsUtils::LayerTestsCommon {
             results.emplace_back(std::make_shared<ngraph::opset3::Result>(output));
         }
 
+        results.insert(results.end(), m_additionalResults.begin(), m_additionalResults.end());
```
Comment (gladilov-gleb) [inline inference-engine/tests/functional/plugin/myriad/subgraph_tests/dsr_tests_common.hpp:74]: Why did you need this?

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 1 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: Fixed.

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 2 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: Good suggestion! Done.

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 3 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: I'm agree, done.

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 5 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: In bidirectional mode the shapes may have different sizes and this broadcast rule is right alignment, so I go from the right to the left. I've used `reverse` just because I've used the `push_back`, but I think it's better to insert in the beggining of the vector, so no `reverse` is needed. Done.

File: inference-engine/tests/functional/plugin/myriad/subgraph_tests/dsr_tests_common.hpp (hunk 6 above)
Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/myriad/subgraph_tests/dsr_tests_common.hpp]: I've taken a look once more and I've found that it's not a good idea to use `DSR_TestsCommon` because I actually doesn't use it's features, so I've changed these classes to be based on `TestsCommon`, so this change is not needed anymore.

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 4 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: Explained below.

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 5 above)
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: Since you push at the beginning instead of back + removed reverse at the same time - does it mean you have now different result from previous version?

input: 0 1 2 3

Previously:

* after for-loop: 3 2 1 0
* after reverse: 0 1 2 3

Now:

* after for-loop: 3 2 1 0

Is it expected?

File: inference-engine/tests/functional/plugin/myriad/subgraph_tests/nonzero_broadcast.cpp [hunk 7]
```diff
@@ -2,70 +2,172 @@
...
-        const auto tensorParam = std::make_shared<ngraph::opset3::Parameter>(
-                tensorType, tensorShape);
-        const auto nonZero = std::make_shared<ngraph::opset3::NonZero>(tensorParam);
-        const auto shapeOfNonZero = std::make_shared<ngraph::opset3::ShapeOf>(nonZero);
+        const auto& upperBoundShape = broadcastParams.targetShape.upperBoundShape;
+        const auto& realShape = broadcastParams.targetShape.shape;
+
+        while (upperBoundShape[m_dynamicAxis] == realShape[m_dynamicAxis]) {
+            m_dynamicAxis++;
+        }
```
Comment (gladilov-gleb) [inline inference-engine/tests/functional/plugin/myriad/subgraph_tests/nonzero_broadcast.cpp:43]: Nitpick: extract in a separate function

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 5 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: No, it's not fully correct. The for-loop itself goes from the end to the begginning of the input.
You're correctly said about the previous version, but now it still goes from the end to the beginning of the input, but each element is inserted into the beginning of the vector, which means that 0 will be processed last and will be first in the result vector, so the output will be the same: 0 1 2 3

Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: Yea, you're right. And as I understand it's impossible to go from left to right and push at the back due to operation specification?

**Review by gladilov-gleb — APPROVED** [review]

File: inference-engine/tests/functional/plugin/myriad/subgraph_tests/nonzero_broadcast.cpp (hunk 7 above)
Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/myriad/subgraph_tests/nonzero_broadcast.cpp]: Thanks! Done.

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp (hunk 5 above)
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: Heh, I don't agree on this. Going from the beginning to the end and push_back is idiomatic and always easier to reason about.

Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp]: OK, fixed.

**Comment by AndrewBakalinIntel** [comment]
@Maxim-Doronin could you please merge?
